coderd/workspaceagents.go

```diff
@@ -1462,6 +1468,72 @@ func (api *API) workspaceAgentsExternalAuthListen(ctx context.Context, rw http.R
 	}
 }
 
+// @Summary Coordinate multiple workspace agents
+// @ID coordinate-multiple-workspace-agents
+// @Security CoderSessionToken
+// @Tags Agents
+// @Success 101
+// @Router /users/me/tailnet [get]
+func (api *API) tailnet(rw http.ResponseWriter, r *http.Request) {
+	ctx := r.Context()
+	apiKey, ok := httpmw.APIKeyOptional(r)
+	if !ok {
+		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+			Message: "Cannot use \"me\" without a valid session.",
+		})
+		return
+	}
+
+	version := "2.0"
+	qv := r.URL.Query().Get("version")
+	if qv != "" {
+		version = qv
+	}
+	if err := proto.CurrentVersion.Validate(version); err != nil {
+		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+			Message: "Unknown or unsupported API version",
+			Validations: []codersdk.ValidationError{
+				{Field: "version", Detail: err.Error()},
+			},
+		})
+		return
+	}
+
+	peerID, err := api.handleResumeToken(ctx, rw, r)
+	if err != nil {
+		// handleResumeToken has already written the response.
+		return
+	}
+
+	api.WebsocketWaitMutex.Lock()
+	api.WebsocketWaitGroup.Add(1)
+	api.WebsocketWaitMutex.Unlock()
+	defer api.WebsocketWaitGroup.Done()
+
+	conn, err := websocket.Accept(rw, r, nil)
+	if err != nil {
+		httpapi.Write(ctx, rw, http.StatusBadRequest, codersdk.Response{
+			Message: "Failed to accept websocket.",
+			Detail:  err.Error(),
+		})
+		return
+	}
+	ctx, wsNetConn := codersdk.WebsocketNetConn(ctx, conn, websocket.MessageBinary)
+	defer wsNetConn.Close()
+	defer conn.Close(websocket.StatusNormalClosure, "")
+
+	go httpapi.Heartbeat(ctx, conn)
+	err = api.TailnetClientService.ServeUserClient(ctx, version, wsNetConn, tailnet.ServeUserClientOptions{
+		PeerID:          peerID,
+		UserID:          apiKey.UserID,
```

Comment:
I realised we can do this RBAC check *during* the DB query, by just inserting the auth-filter -  so I've added that back in the below PR. Currently, we'll send updates owned by the given user (whose ID is provided in the stream request) for all workspaces that we can `ActionRead`.

For authorizing tunnel creation, I've currently just got it querying the DB to fetch the full `database.Workspace`, and then using that to construct an RBAC object. 
It could be better to just store enough state in `Subscription` to form an RBAC object, but plumbing one or more current `Subscription`s through to the tunnel Authorize call seems horribly janky. If we store the subscriptions in a map in the `WorkspaceUpdatesProvider`, it's a lot easier to inspect the state.

On the other hand, this seems equally as brittle, if we change the columns needed to form a workspace RBAC object (e.g. we added ACL lists, a `shared_with` column, etc), we'd need to update our query, and our subscription state. Perhaps we just only send workspace updates for those you can `ActionSSH`?

Reply:
I think it's fine to hit the DB for authorization, but we might need to rate limit: I don't want a misbehaving client DOS'ing the whole Coder deployment.  In the ordinary case, you'll only be adding tunnels on startup of CoderVPN or when you start a new workspace, which is infrequent (a few queries per day per Coder Desktop client).